node/circuits.js

```diff
@@ -0,0 +1,154 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+"use strict";
+
+var States = require('./states');
+var StateMachine = require('./state_machine');
+
+// The each circuit uses the circuits collection as the "nextHandler" for
+// "shouldRequest" to consult.  Peers use this hook to weight peers both by
+// healthy and other factors, but the circuit only needs to know about health
+// before forwarding.
+
+function AlwaysShouldRequestHandler() { }
+
+AlwaysShouldRequestHandler.prototype.shouldRequest = function shouldRequest() {
+    return true;
+};
+
+var alwaysShouldRequestHandler = new AlwaysShouldRequestHandler();
+
+function Circuits(options) {
+    var self = this;
+    self.circuits = {};
+    self.circuitsByServiceName = {};
+    self.stateOptions = {
+        stateMachine: self,
+        nextHandler: alwaysShouldRequestHandler,
+        timers: options.timers,
+        random: options.random,
+        period: options.period,
+        maxErrorRate: options.maxErrorRate,
+        minRequests: options.minRequests,
+        probation: options.probation,
+    };
+    self.shouldRequestOptions = {};
+}
+
+Circuits.prototype.getCircuit = function getCircuit(callerName, serviceName, endpointName) {
+    var self = this;
+    // TODO escape components
+    var circuitName = callerName + '::' + serviceName + '::' + endpointName;
+    var circuit = self.circuits[circuitName];
+    if (!circuit) {
+        circuit = new Circuit();
+        circuit.circuitName = circuitName;
+        circuit.callerName = callerName;
+        circuit.serviceName = serviceName;
+        circuit.endpointName = endpointName;
+        circuit.shouldRequestOptions = self.shouldRequestOptions;
+        circuit.stateOptions = self.stateOptions;
+        self.circuits[circuitName] = circuit;
+    }
+    return circuit;
+};
+
+Circuits.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
+    var self = this;
+    var callerName = req.headers.cn;
+    if (!callerName) {
+        return buildRes().sendError('BadRequest', 'All requests must have the cn (caller service name) header');
```

Comment:
This is a breaking change. This cannot be merged until we transition people over to having the cn header.